cunumeric/module.py

```diff
@@ -3506,6 +3506,46 @@ def put(
     a.put(indices=indices, values=values, mode=mode)
 
 
+@add_boilerplate("a", "mask", "values")
+def putmask(a: ndarray, mask: ndarray, values: ndarray) -> None:
+    """
+    putmask(a, mask, values)
+    Changes elements of an array based on conditional and input values.
+    Sets ``a.flat[n] = values[n]`` for each n where ``mask.flat[n]==True``.
+    If `values` is not the same size as `a` and `mask` then it will repeat.
+    This gives behavior different from ``a[mask] = values``.
+
+    Parameters
+    ----------
+    a : ndarray
+        Target array.
+    mask : array_like
+        Boolean mask array. It has to be the same shape as `a`.
+    values : array_like
+        Values to put into `a` where `mask` is True. If `values` is smaller
+        than `a` it will be repeated.
+
+    See Also
+    --------
+    numpy.putmask
+
+    Availability
+    ------------
+    Multiple GPUs, Multiple CPUs
+    """
+    if not a.shape == mask.shape:
+        raise ValueError("mask and data must be the same size")
+
+    mask = mask._warn_and_convert(np.dtype(bool))
+
+    if a.dtype != values.dtype:
+        values = values._warn_and_convert(a.dtype)
+    if values.shape != a.shape and values.size != 1:
```

Comment:
the first condition should be `values.size != size`. Here's an example you can handle without wrapping or tiling but simply using numpy broadcasting:

```
import numpy as np
a = np.full((3, 3), 3)
np.putmask(a, np.full((3, 3), True), np.full(3, 10))
```

Reply:
unless stated otherwise, it's safe to assume that all related array arguments follow the numpy broadcasting semantics.

Reply:
I adding the check if arrays can be broadcasted. We will cal `wrap` only for the case when they can't